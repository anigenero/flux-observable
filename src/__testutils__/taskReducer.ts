import {Reducer} from 'react';
import {from, of} from 'rxjs';
import {switchMap} from 'rxjs/operators';
import {Epic} from '../combine.epics';
import {ofType} from '../ofType';

export type Dependencies = {
    service: (task: string) => string;
};

export namespace TaskActionTypes {
    export const SET_TASK = 'set_task';

    export const START_TASK = 'start_task';

}

export type TaskState = {
    task: string;
};

export type TaskActions = {
    [TaskActionTypes.SET_TASK]: {
        type: typeof TaskActionTypes.SET_TASK;
        task: string;
    };

    [TaskActionTypes.START_TASK]: {
        type: typeof TaskActionTypes.START_TASK;
        task: string;
    };

};

export type TaskReducerAction =
    TaskActions[typeof TaskActionTypes.SET_TASK];

export type TaskAction =
    TaskReducerAction |
    TaskActions[typeof TaskActionTypes.START_TASK];

export namespace TaskDispatchActions {

    export const setTask = (task: string): TaskActions[typeof TaskActionTypes.SET_TASK] => ({
        type: TaskActionTypes.SET_TASK,
        task
    });

    export const startTask = (task: string): TaskActions[typeof TaskActionTypes.START_TASK] => ({
        type: TaskActionTypes.START_TASK,
        task
    });

}

export const defaultState: TaskState = {
    task: null
};

export const taskReducer: Reducer<TaskState, TaskReducerAction> = (prevState, action) => {
    switch (action.type) {

        case TaskActionTypes.SET_TASK: {
            return {
                task: action.task
            };
        }

        default: {
            return prevState;
        }

    }
};

export namespace ReducerEpics {

    export const startTask: Epic<TaskAction, TaskAction, TaskState, Dependencies> =
        (action$, state$, {service}) =>
            action$.pipe(
                ofType(TaskActionTypes.START_TASK),
                switchMap(({task}) =>
                    from(service(task)).pipe(
                        switchMap((result) => of(TaskDispatchActions.setTask(result)))
                    )
                )
            );

}
