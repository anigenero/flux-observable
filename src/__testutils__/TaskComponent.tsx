import React, {Dispatch, FunctionComponent} from 'react';
import {TaskAction, TaskState} from './taskReducer';

export type TaskComponentProps = {
    useContext: () => {
        dispatch: Dispatch<TaskAction>;
        state: TaskState;
    };
};

export const TaskComponent: FunctionComponent<TaskComponentProps> = ({useContext}) => {

    const {state, dispatch} = useContext();

    return (
        <div>
            {state.task}
        </div>
    );

};
