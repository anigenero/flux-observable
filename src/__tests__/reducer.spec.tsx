import React from 'react';
import {create} from 'react-test-renderer';
import {TaskComponent} from '../__testutils__/TaskComponent';
import {defaultState, ReducerEpics, taskReducer} from '../__testutils__/taskReducer';
import {createObservableReducerContext} from '../index';

describe('createObservableReducerContext', () => {

    const serviceFunction = jest.fn().mockResolvedValue('task1');

    it('should create context', () => {

        const {Provider, useObservableContext} = createObservableReducerContext(taskReducer, [
            ReducerEpics.startTask
        ], {service: serviceFunction});

        const tree = create(
            <Provider options={{state: defaultState}}>
                <TaskComponent useContext={useObservableContext}/>
            </Provider>
        ).toJSON();

        expect(tree).toMatchSnapshot();

    });

});
