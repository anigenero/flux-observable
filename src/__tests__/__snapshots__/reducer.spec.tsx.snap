// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createObservableReducerContext should create context 1`] = `<div />`;
